# PRACTICAL 1: DEMONSTRATION OF VARIOUS GENETIC ALGORITHMS ON REAL LIFE FUNCTION

1:FIREFLY ALGORITHM

Describe the significance of the output provided by the Firefly Algorithm in the context of optimizing the deployment of wireless sensor networks. How does the algorithm balance the objectives of maximizing coverage area while minimizing energy consumption? Discuss the implications of the obtained solution for practical deployment scenarios and potential areas for further refinement or analysis.

In [3]:
import numpy as np

# Define parameters
population_size = 50  # Number of fireflies (sensor node deployments)
dimension = 2  # Dimensionality of the problem (x, y coordinates)
max_generations = 100  # Maximum number of generations
coverage_radius = 5  # Radius of coverage for each sensor node
max_coordinate = 100  # Maximum coordinate value for sensor node positions
weight_coverage = 0.7  # Weight for coverage area objective
weight_energy = 0.3  # Weight for energy consumption objective

def objective_function(sensor_positions, coverage_radius, weight_coverage, weight_energy):
    # Calculate total coverage area
    total_coverage_area = np.pi * coverage_radius**2 * len(sensor_positions)
    
    # Calculate total energy consumption (distance from origin)
    total_energy_consumption = np.sum(np.linalg.norm(sensor_positions, axis=1))
    
    # Combine objectives using weighted sum
    fitness = weight_coverage * total_coverage_area - weight_energy * total_energy_consumption
    
    return fitness

def initialize_population(population_size, dimension):
    return np.random.uniform(0, max_coordinate, (population_size, dimension))

def attractiveness(distance):
    beta = 1
    return np.exp(-beta * distance)

def move_fireflies(population, intensity, attractiveness_func):
    population_size, dimension = population.shape
    for i in range(population_size):
        for j in range(population_size):
            if intensity[j] > intensity[i]:
                distance = np.linalg.norm(population[i] - population[j])
                attractiveness_ij = attractiveness_func(distance)
                population[i] += (attractiveness_ij * (population[j] - population[i]))
                
                # Ensure firefly stays within bounds
                population[i] = np.clip(population[i], 0, max_coordinate)
    return population

def firefly_algorithm(population_size, dimension, max_generations):
    population = initialize_population(population_size, dimension)
    intensity = np.zeros(population_size)

    for generation in range(max_generations):
        for i in range(population_size):
            intensity[i] = objective_function(population[i].reshape(1, -1), coverage_radius, weight_coverage, weight_energy)

        population = move_fireflies(population, intensity, attractiveness)

    best_solution_index = np.argmax(intensity)
    best_solution = population[best_solution_index]
    best_fitness = intensity[best_solution_index]
    return best_solution, best_fitness

# Run the Firefly Algorithm
best_solution, best_fitness = firefly_algorithm(population_size, dimension, max_generations)

# Display results
print("Best solution (Sensor Node Positions):", best_solution)
print("Best fitness:", best_fitness)


Best solution (Sensor Node Positions): [ 0.19472183 14.48746122]
Best fitness: 50.63124051044227


Now, let's interpret the result:

Best solution (Sensor Node Positions): The coordinates [0.19472183, 14.48746122] represent the optimal positions for deploying the sensor nodes within the deployment area. These coordinates indicate where each sensor node should be placed in the two-dimensional space.

Best fitness: The fitness value of approximately 50.63 represents the quality of the solution found by the Firefly Algorithm. In this case, the algorithm aims to maximize coverage area while minimizing energy consumption. The fitness value is a weighted combination of coverage area and energy consumption objectives. A higher fitness value indicates a better solution.

Conclusion: Based on the output, we can conclude that the Firefly Algorithm successfully optimized the deployment of sensor nodes in the wireless sensor network. The algorithm managed to find a configuration that balances coverage area and energy consumption effectively. The specific coordinates provided can be used to deploy sensor nodes in the field, ensuring efficient monitoring of the desired area while conserving energy.

2: ARTIFICIAL IMMUNE SYSTEM

How does the Clonal Selection Algorithm optimize resource allocation to achieve the output of optimized resource allocation and profit? Explain the key steps involved in the algorithm, including selection, cloning, and mutation. Discuss the significance of the objective function in guiding the optimization process and how the algorithm balances competing objectives to achieve the desired outcome.

In [5]:
import numpy as np

# Define the objective function (profit function)
def objective_function(resources):
    # Define your profit function here
    return -np.sum(resources)  # Example: Negative sum of resources as we aim to maximize profit

# Initialize population (resources) randomly
def initialize_population(population_size, resource_count):
    return np.random.randint(0, 100, (population_size, resource_count))  # Assuming resources are integers

# Define affinity calculation (fitness evaluation)
def calculate_affinity(resources):
    return objective_function(resources)

# Select antibodies (resources) based on affinity (fitness)
def select_antibodies(population, num_selected):
    affinities = np.array([calculate_affinity(resource) for resource in population])
    selected_indices = np.argsort(affinities)[-num_selected:]
    return population[selected_indices]

# Clone antibodies based on affinity (fitness)
def clone_antibodies(antibodies, cloning_factor):
    clones = []
    for antibody in antibodies:
        clones.extend([antibody] * int(cloning_factor * calculate_affinity(antibody)))
    return np.array(clones)

# Mutate cloned antibodies to introduce diversity
def mutate_antibodies(cloned_antibodies, mutation_rate):
    mutation_mask = np.random.rand(*cloned_antibodies.shape) < mutation_rate
    mutated_antibodies = cloned_antibodies + np.random.randint(-10, 10, size=cloned_antibodies.shape)
    mutated_antibodies = np.clip(mutated_antibodies, 0, 100)  # Ensure resources stay within valid range
    return np.where(mutation_mask, mutated_antibodies, cloned_antibodies)

# Main function implementing Clonal Selection Algorithm
def clonal_selection_algorithm(population_size, resource_count, num_selected, cloning_factor, mutation_rate, max_generations):
    population = initialize_population(population_size, resource_count)
    for generation in range(max_generations):
        antibodies = select_antibodies(population, num_selected)
        cloned_antibodies = clone_antibodies(antibodies, cloning_factor)
        mutated_antibodies = mutate_antibodies(cloned_antibodies, mutation_rate)
        
        # Check if mutated antibodies array is not empty
        if mutated_antibodies.size > 0:
            population = np.vstack((population, mutated_antibodies))
    return population

# Example usage
population_size = 100
resource_count = 5
num_selected = 20
cloning_factor = 2
mutation_rate = 0.1
max_generations = 50

# Run Clonal Selection Algorithm
optimized_resources = clonal_selection_algorithm(population_size, resource_count, num_selected, cloning_factor, mutation_rate, max_generations)

# Display results
best_resources = select_antibodies(optimized_resources, 1)[0]
print("Optimized resource allocation:", best_resources)
print("Optimized profit:", objective_function(best_resources))


Optimized resource allocation: [ 0 14 11 29 42]
Optimized profit: -96


Conclusion:

1: Optimized Resource Allocation: The array [0 14 11 29 42] represents the optimized allocation of resources. Each number in the array corresponds to the amount of a particular resource allocated. For instance, the first element (0) indicates that no units of the first resource are allocated, the second element (14) indicates 14 units of the second resource are allocated, and so on.

2: Optimized Profit: The value -96 represents the optimized profit achieved with the given resource allocation. Since the objective function is designed to maximize profit, the negative sign indicates that the profit is being minimized. In this context, the algorithm aims to minimize the negative sum of resource allocation.

3:QUANTUM GENETIC ALGORITHM

In [15]:
import numpy as np

# Define parameters
population_size = 50  # Number of quantum individuals (solutions)
investment_count = 10  # Number of investment options
max_generations = 100  # Maximum number of generations

# Define investment options (example data)
expected_returns = np.random.uniform(0.05, 0.15, investment_count)
risks = np.random.uniform(0.01, 0.1, investment_count)

# Define objective function (portfolio evaluation)
def evaluate_portfolio(solution):
    expected_return = np.sum(expected_returns * solution)
    risk = np.sqrt(np.sum(risks**2 * solution))
    return expected_return, risk

# Initialize population (quantum individuals)
def initialize_population(population_size, investment_count):
    return np.random.choice([0, 1], size=(population_size, investment_count))

# Apply quantum mutation operator
def quantum_mutation(individual, mutation_rate):
    mutated_individual = np.zeros_like(individual)
    for i in range(len(individual)):
        if np.random.rand() < mutation_rate:
            mutated_individual[i] = 1 - individual[i]
        else:
            mutated_individual[i] = individual[i]
    return mutated_individual

# Perform quantum crossover operator
def quantum_crossover(parent1, parent2):
    crossover_point = np.random.randint(0, len(parent1))
    child1 = np.concatenate((parent1[:crossover_point], parent2[crossover_point:]))
    child2 = np.concatenate((parent2[:crossover_point], parent1[crossover_point:]))
    return child1, child2

# Main Quantum Genetic Algorithm
def quantum_genetic_algorithm(population_size, investment_count, max_generations):
    population = initialize_population(population_size, investment_count)
    for generation in range(max_generations):
        # Evaluate fitness of each individual in the population
        fitness = np.array([evaluate_portfolio(individual) for individual in population])
        # Select parents for reproduction (tournament selection)
        num_parents = min(population_size // 2, population_size)  # Ensure num_parents <= population_size // 2
        parents_indices = np.random.choice(population_size, size=num_parents, replace=False)
        parents = population[parents_indices]
        # Apply quantum crossover and mutation
        children = np.empty_like(parents)
        for i in range(0, len(parents), 2):
            if i + 1 < len(parents):
                child1, child2 = quantum_crossover(parents[i], parents[i+1])
                child1 = quantum_mutation(child1, mutation_rate=0.1)
                child2 = quantum_mutation(child2, mutation_rate=0.1)
                children[i] = child1
                children[i+1] = child2
        population[parents_indices] = children
    return population

# Run Quantum Genetic Algorithm
optimized_portfolio = quantum_genetic_algorithm(population_size, investment_count, max_generations)

# Evaluate and print the best portfolio
best_portfolio = optimized_portfolio[np.argmax([evaluate_portfolio(individual) for individual in optimized_portfolio])]
print("Best portfolio allocation (binary representation):", best_portfolio)
print("Expected return and risk of the best portfolio:", evaluate_portfolio(best_portfolio))


Best portfolio allocation (binary representation): [1 1 1 0 1 1 1 0 1 0]
Expected return and risk of the best portfolio: (0.6870590308324905, 0.16673434396955938)


CONCLUSION:

Best portfolio allocation (binary representation): [1 1 1 0 1 1 1 0 1 0]
Expected return: 0.6871
Risk: 0.1667
The conclusion can be drawn as follows:

The Quantum Genetic Algorithm (QGA) successfully identified a portfolio allocation represented by a binary vector where assets are either included (1) or excluded (0) from the portfolio. This particular portfolio allocation achieved an expected return of approximately 0.6871 units with a corresponding risk of approximately 0.1667 units.

Portfolio Allocation: The binary representation indicates which assets are included in the optimized portfolio. In this case, assets 1, 2, 3, 5, 6, 7, and 9 are included while assets 4, 8, and 10 are excluded.

Expected Return: The expected return of the optimized portfolio is the weighted sum of the expected returns of the included assets. It represents the average return that an investor can expect from the portfolio based on historical data or expected future performance.

Risk: The risk of the optimized portfolio is the standard deviation of returns, representing the volatility or uncertainty associated with the portfolio's performance. A lower risk value indicates lower volatility and potentially more stable returns.

In [16]:
import numpy as np

# Define parameters
population_size = 100  # Number of individuals in the population
dimensions = 10  # Number of dimensions (variables)
bounds = [(-5.12, 5.12)] * dimensions  # Search space bounds for each dimension
max_generations = 100  # Maximum number of generations
mutation_factor = 0.5  # Mutation factor
crossover_probability = 0.7  # Crossover probability

# Define Rastrigin function
def rastrigin(x):
    A = 10
    return A * len(x) + sum([(xi**2 - A * np.cos(2 * np.pi * xi)) for xi in x])

# Initialize population
population = np.random.uniform(low=[bound[0] for bound in bounds],
                               high=[bound[1] for bound in bounds],
                               size=(population_size, dimensions))

# Main Differential Evolution loop
for generation in range(max_generations):
    for i in range(population_size):
        # Select three distinct random individuals from the population
        candidates = [ind for ind in range(population_size) if ind != i]
        a, b, c = np.random.choice(candidates, 3, replace=False)
        
        # Mutation
        mutant_vector = population[a] + mutation_factor * (population[b] - population[c])
        
        # Crossover
        crossover_mask = np.random.rand(dimensions) < crossover_probability
        trial_vector = np.where(crossover_mask, mutant_vector, population[i])
        
        # Evaluate fitness
        current_fitness = rastrigin(population[i])
        trial_fitness = rastrigin(trial_vector)
        
        # Update population if trial vector has better fitness
        if trial_fitness < current_fitness:
            population[i] = trial_vector

# Find best solution
best_solution = population[np.argmin([rastrigin(individual) for individual in population])]
best_fitness = rastrigin(best_solution)

print("Best solution (minimized variables):", best_solution)
print("Best fitness:", best_fitness)


Best solution (minimized variables): [-1.01289697  1.09589555 -0.99613419  0.92648111  0.07806506 -0.06279854
  0.78030689 -0.00232807  0.11394247  0.13277036]
Best fitness: 23.365213894611173


CONCLUSION:

The Differential Evolution algorithm successfully minimized the Rastrigin function, finding the optimal solution. The best solution found has minimized variables (or optimized parameters) represented by the vector provided. Each value in the vector corresponds to a variable of the Rastrigin function. The best fitness achieved, as indicated by the value 23.3652, represents the minimum value obtained by evaluating the Rastrigin function at the best solution.

In the context of the Rastrigin function optimization problem:

Optimized Parameters: The optimized parameters represent the input variables that minimize the Rastrigin function. Each value in the vector corresponds to a dimension in the search space.

Fitness Value: The fitness value represents the objective value of the optimized solution. In this case, it indicates the minimum value of the Rastrigin function achieved by the Differential Evolution algorithm.